### PR TITLE
Adding ability to redirect stdio to filedescriptor

### PR DESCRIPTION
**Pull request by robinbernon**
Adding ability to redirect stdio to filedescriptor

### CODE REVIEW

File: filedescriptor/src/windows.rs
Comment (wez) [inline filedescriptor/src/windows.rs:218]: I think you want to `FileDescriptor::dup(std_original)` here so that you have a managed duplicate handle.  It may be slightly better to `FileDescriptor::new(std_original)` so that you take complete ownership of that handle; that is semantically closer to what this would do on unix.

File: filedescriptor/src/windows.rs
Comment (wez) [inline filedescriptor/src/windows.rs:219]: I think this should assign a duplicate of `self.handle`, otherwise when `self` is dropped, we break the output for the whole program.

Really, I think this should be a "class" function and not accept `self` but instead be passed something that is `AsRawFileDescriptor`.  Then this logic would be to `FileDescriptor::dup()` the passed handle, and if `SetStdHandle` succeeds, that duplicated handle would have `into_raw_descriptor` called on it to transfer ownership to the OS.

File: filedescriptor/src/windows.rs
Comment (robinbernon) [reply filedescriptor/src/windows.rs]: Have made this change to the code now

File: filedescriptor/src/windows.rs
Comment (robinbernon) [reply filedescriptor/src/windows.rs]: Have made this change to the code now

File: filedescriptor/src/unix.rs
Comment (wez) [inline filedescriptor/src/unix.rs:283]: I was envisioning passing `f` directly to `dup2` and avoiding the intermediate `cloned_handle`:

```suggestion
        unsafe { FileDescriptor::dup2(f, libc::STDIN_FILENO) }?;
```

File: filedescriptor/src/windows.rs
Comment (wez) [inline filedescriptor/src/windows.rs:277]: I think this should transfer ownership to the OS, otherwise the handle will get closed when `cloned_handle` is dropped when this function returns:

```suggestion
                cloned_handle.into_raw_handle() as *mut winapi::ctypes::c_void,
```

File: filedescriptor/src/unix.rs
Comment (wez) [inline filedescriptor/src/unix.rs:288]: FWIW, as a matter of style, I was envisioning this being a single function like this:

```rust
pub(crate) fn redirect_stdio_impl<F: AsRawFileDescriptor>(
        f: &F,
        stdio: StdioDescriptor,
    ) -> anyhow::Result<Self> {
```

that internally did:

```rust
let std_original = FileDescriptor::dup(match stdio {
   StdioDescriptor::Stdin => libc::STDIN_FILENO,
   StdioDescriptor::Stdout => libc::STDOUT_FILENO,
   StdioDescriptor::Stderr => libc::STDERR_FILENO,
))?;
```

and similarly for the Windows implementation

File: filedescriptor/src/unix.rs
Comment (wez) [inline filedescriptor/src/unix.rs:169]: For a bit of complexity, I think `dup3` is a linux only interface, so I suspect that you'll need to do something like:

```rust

#[cfg(target_os = "linux")]
{
   // try to use dup3 here
}
// fall back to non-atomic dup2 here
```

File: filedescriptor/src/lib.rs
Comment (wez) [inline filedescriptor/src/lib.rs:272]: nit: the case of these doesn't match those in the enum.  I'd be fine omitting this line TBH, but if we keep it, I'd suggest making this consistent! :)

File: filedescriptor/src/unix.rs
Comment (robinbernon) [reply filedescriptor/src/unix.rs:288]: Ah ok sure. Just thought the execution would be slightly faster with what I've done currently as it avoids having to move the enum into the interior methods. Will change this though to do it this way instead then 👍

File: filedescriptor/src/unix.rs
Comment (robinbernon) [reply filedescriptor/src/unix.rs:169]: It's already in the unix mod. This is cfg toggled in lib so should never get called when running in windows.

File: filedescriptor/src/lib.rs
Comment (robinbernon) [reply filedescriptor/src/lib.rs:272]: Sure, makes sense :) will make these all lowercase then.

File: filedescriptor/src/unix.rs
Comment (wez) [reply filedescriptor/src/unix.rs:288]: FWIW, that enum is super cheap to move around because it fits in a register.  Also: the runtime cost of this code path will be dominated by the cost of the file descriptor related syscalls, so even if the enum was big it still wouldn't matter.

File: filedescriptor/src/unix.rs
Comment (robinbernon) [reply filedescriptor/src/unix.rs:169]: If you want me to build a windows version I'm not sure how to deal with cloexec management. What would I need to do for this?

Comment (wez) [reply filedescriptor/src/unix.rs:169]: what I meant was, the only unix with `dup3` is linux.  macos and freebsd don't have it, so you'll need to explicitly check for linux to call `dup3`.

File: filedescriptor/src/unix.rs
Comment (robinbernon) [reply filedescriptor/src/unix.rs]: Done

File: filedescriptor/src/windows.rs
Comment (robinbernon) [reply filedescriptor/src/windows.rs]: Done

File: filedescriptor/src/unix.rs
Comment (robinbernon) [reply filedescriptor/src/unix.rs]: Cool, have now made the change

File: filedescriptor/src/unix.rs
Comment (robinbernon) [reply filedescriptor/src/unix.rs:170]: Ah apologies, completely misread this. Will change now.

File: filedescriptor/src/lib.rs
Comment (robinbernon) [reply filedescriptor/src/lib.rs]: Done

File: filedescriptor/src/unix.rs
Comment (robinbernon) [reply filedescriptor/src/unix.rs:170]: Done

File: filedescriptor/src/unix.rs
Comment (wez) [inline filedescriptor/src/unix.rs:296]: I think this may also need `.into_raw_fd()` here to avoid closing the fd when this function returns.

It would be great to add a regression test for this :)

File: filedescriptor/src/unix.rs
Comment (wez) [inline filedescriptor/src/unix.rs:179]: this line has to be enclosed in `#[cfg(target_os = "linux")]` block in order to compile+link on !linux systems

Comment (robinbernon) [reply filedescriptor/src/unix.rs:179]: Sorry not sure what you mean here. As is, this forces dup2 for non linux, otherwise, if linux, then it proceeds to attempt dup3 first. Is this not what you want here?

File: filedescriptor/src/unix.rs
Comment (robinbernon) [reply filedescriptor/src/unix.rs]: I've made the change to add ```.into_raw_fd()``` into the code. Not sure what do make for the regression test. I had a plan instead to implement this now in a branch in the gag crate. Was gonna run my tests for these through the gag create instead.

File: filedescriptor/src/unix.rs
Comment (wez) [reply filedescriptor/src/unix.rs:179]: Like I sketched out in the original comment; you need to enclose the block of code that calls dup3 in the cfg check, then under it, fall back to the non atomic dup2.

The issue to resolve is that you can only reference `dup3` on linux.  The program won't compile on any other platform unless dup3 is in a linux only block.

Comment (robinbernon) [reply filedescriptor/src/unix.rs:179]: The code above already does what you want here => if not linux then skip dup3 and do only dup2

Comment (robinbernon) [reply filedescriptor/src/unix.rs:179]: image

Comment (wez) [reply filedescriptor/src/unix.rs:179]: The intent of your change is correct, but the way it is expressed won't compile on mac or freebsd because `dup3` doesn't exist there.  You need to ensure that `dup3` is only referenced on linux, but right now it is referenced on all unix systems.

File: filedescriptor/src/unix.rs
Comment (wez) [inline filedescriptor/src/unix.rs:296]: This doesn't compile because all we know about `F` is that it is `AsRawFileDescriptor`.  That means that we can call `as_raw_file_descriptor` to borrow (but not take ownership of) the `f` parameter.

```suggestion
        unsafe { FileDescriptor::dup2(f.as_raw_file_descriptor(), std_descriptor) }?;
```_

Comment (wez) [reply filedescriptor/src/unix.rs:296]: Actually, I think we just want to pass this through:

```suggestion
        unsafe { FileDescriptor::dup2(f, std_descriptor) }?;
```

Comment (robinbernon) [reply filedescriptor/src/unix.rs]: Done

File: filedescriptor/src/unix.rs
Comment (robinbernon) [reply filedescriptor/src/unix.rs:180]: Done

Comment (wez) [reply filedescriptor/src/unix.rs:180]: GH actions need approval on each change, which is a bit of a drag, however, the freebsd CI is showing that this doesn't compile; it you structure it by enclosing the appropriate code in a block as I said in the original comment, you'll have the most success.

You can keep iterating against that until it goes green and then I can approve and run the rest of the platforms!

**Review by wez — CHANGES_REQUESTED** [review]
Please fixup the freebsd build per earlier comments; thanks!

File: filedescriptor/src/unix.rs
Comment (robinbernon) [reply filedescriptor/src/unix.rs]: @wez could you run the pipeline again please?

**Comment by robinbernon** [comment]
> Please fixup the freebsd build per earlier comments; thanks!

Think it's fixed now. Can you please re-run the pipeline?

**Comment by robinbernon** [comment]
@wez is there more info on why freebsd-12 failed here?

**Comment by robinbernon** [comment]
@wez looks like it's not even getting to the  build stage on freebsd-12, is there something wrong with the environment setup here?

**Comment by wez** [comment]
> @wez looks like it's not even getting to the build stage on freebsd-12, is there something wrong with the environment setup here?

Appears to be: https://github.com/rust-lang/rustup/issues/2774

Let's see if https://github.com/wez/wezterm/commit/caed905626123cf5cec1368377330413a3e1ab1a resolves it; you can try rebasing the PR on that commit.

File: filedescriptor/src/unix.rs
Comment (wez) [inline filedescriptor/src/unix.rs:200]: I think this can match the code in `non_atomic_dup2` above:

```suggestion
                    handle_type: (),
```

File: filedescriptor/src/unix.rs
Comment (wez) [inline filedescriptor/src/unix.rs:300]:
```suggestion
        unsafe { FileDescriptor::dup2(f, std_descriptor) }?.into_raw_fd();
```

File: filedescriptor/src/windows.rs
Comment (wez) [inline filedescriptor/src/windows.rs:282]: I think you can simplify this:

```suggestion
                cloned_handle.into_raw_handle() as *mut _,
```

Comment (robinbernon) [reply filedescriptor/src/windows.rs]: Done

File: filedescriptor/src/unix.rs
Comment (robinbernon) [reply filedescriptor/src/unix.rs]: Done

File: filedescriptor/src/unix.rs
Comment (robinbernon) [reply filedescriptor/src/unix.rs]: Done

**Comment by robinbernon** [comment]
@wez one more run please?

**Comment by robinbernon** [comment]
Guess I must have forgotten to fmt. Have done this now soz. Could you please try again?

**Comment by wez** [comment]
I think you forgot to push :)

**Pull request closed without merging**
